### ISSUE TITLE
[Bug] `beartype.door` permissively accepts unsubscripted `typing.Optional` and `typing.Union` type hint factories as akin to typing Gods!? 😱

### ISSUE THREAD

**Issue by sylvorg** [beartype/beartype#376]
Hello! Guess who's back! ... I'm not doing that joke again, you can't make me! 😤

Get ready for a deluge of replies, starting with this issue! ... At least I think it's an issue!

Is everything supposed to match a `Union` generic when using `beartype.door.is_bearable` and `beartype.door.is_subhint`? For example, with the following block:

```python
from beartype.door import is_bearable, is_subhint
from collections.abc import Callable, Iterable
from typing import Literal, Union

for t in (
    str | int,
    Union,
    Union[str, int],
    Callable,
    Callable[..., ...],
    Iterable,
    Iterable[str],
    Literal,
    Literal[1, 2],
):
    assert is_bearable(t, Union)
    assert is_subhint(t, Union)
```

No `AssertionError` occurs. I'd really like to switch to `is_subhint`, because then I wouldn't have to worry about `issubclass` erroring out on annotations like `Union[...]`, etc., and using `typing.get_origin` all the time! 😅

Thank you kindly for the help!

***

P.S.: I passed the graphics course! 😻 I got a `D`, but still. A win is a win. 😹

**Comment by leycec** [comment]
**Yay!!!** It's my favourite silver Borg, @sylvorg. He's right about everything, of course. @beartype should *definitely* be rejecting unsubscripted type hint factories that otherwise have *no* semantic meaning like:

* `typing.Literal`. Think we nailed this one recently? I dimly recall nailing this. Stay down when I put you down, bugs!
* `typing.Optional`. :cry:
* `typing.Union`. :sob:
* *Probably others.*

Who cares about those, though! We've got an unsubscripted `Union` to crush.

> using `typing.get_origin` all the time!

![](https://media1.tenor.com/m/aKmjR2yVTk0AAAAC/liar-liar-comedy.gif)

> P.S.: I passed the graphics course! 😻 I got a D, but still. A win is a win. 😹

![](https://media1.tenor.com/m/G6wowW9nEeoAAAAC/go-buckeyes.gif)

**Comment by sylvorg** [comment]
> Yay!!! It's my favourite silver Borg, @sylvorg.

Technically, [Sylveon](https://bulbapedia.bulbagarden.net/wiki/Sylveon_(Pok%C3%A9mon)) plus:
![borg](https://github.com/beartype/beartype/assets/94809170/7e44f434-8bf4-406c-b841-aba15995d5b3)

But yes. 😹

This issue is quite exciting because it could reduce quite a bit of code on my end! 😻

**Comment by leycec** [comment]
**Years pass...** before @leycec belatedly realises that @beartype is actually behaving as expected here. Although insane, @beartypes's current behaviour is "working as intended." <sup>\*</sup> 

<sup>\* The terms "working" and "intended" are both nebulous and ill-defined.</sup>

You are now thinking:

> "WHAT YOU SAY!?!?"

Yes. Unsubscripted type hint factories like `typing.Optional` and `typing.Union` are themselves valid type hints. I know. Look. This isn't up to @beartype. This is according to the insane standards. Specifically, PEP 484, which stipulates that unsubscripted type hint factories are **implicitly subscripted** by `typing.Any`. Ergo:

* `typing.Union` implicitly expands to `typing.Union[typing.Any]`. However, PEP 484 also stipulates that a union of one type semantically reduces to only that type! Since this implies that `typing.Union[typing.Any]` semantically reduces to merely `typing.Any`, the unsubscripted type factory `typing.Union` reduces to `typing.Any` as well.
* Similar logic applies to `typing.Optional`, which behaves similarly.

Don't shoot the QA messenger. I don't make the QA-destroying `typing` standards. I only ineffectually complain about the QA-destroying `typing` standards live on GitHub.

2025: *ain't off to a good start is it, folks?* :sob:

**Comment by sylvorg** [comment]
> Oh no.

— Baymax

Looks like I've got some code to revise in my own project as well, then. 😹 I'll have to allow everything to match `typing.Union`, then! Thanks for the heads-up! 😸

**Comment by sylvorg** [comment]
Although... Apparently you can't do something like `typing.Callable[[typing.Union], typing.Union]`, apparently...?

**Comment by leycec** [comment]
**lolbro.** Baymax intensifies and then feels deflated. <sup>*...get it, "deflated"? you see, that was a pun because... oh, nmind.*</sup> 

`typing` sure is super-weird, huh? You have now noticed that deprecated PEP 484-compliant `typing` type hint factories behave *nothing* like PEP 585-compliant builtin type hint factories. Behold! True horror!

```python
>>> from typing import List, Union
>>> list[Union]
list[Union]  # <-- guess that's okay, i guess? thanks, pep 585!
>>> List[Union]
...
TypeError: Plain typing.Union is not valid as type argument
# ^-- thanks fer nuthin', pep 484.
```

PEP 585 and 484 disagree about everything. PEP 585 takes precedence, because PEP 585 destroyed PEP 484. I don't know. Neither does anyone else. It's *probably* best to assume PEP 585 has the right of it – especially because that assumption simplifies everything and makes everything orthogonal.

Basically:

* *Any* type hint factory can be unsubscripted.
* When unsubscripted, a type hint factory defaults to subscription by the `typing.Any` singleton.

Sadly, this was never actually formalized anywhere. PEP 484 strongly hints <sup>...*get it? i'll stop now*</sup> that this is the case, but never comes right out and says it. Because it's easier to be permissive, @beartype assumes that this is the case.

Thus ends the tale of the 5th QA battle. **PEP 585: You win the day!** :flags:

**Comment by sylvorg** [comment]
Fair enough! 😹 I still wish they included the errors regarding how many subscriptions are possible for each type, like unlimited for `tuple` and `Union`, 1 for other iterables like `list`, `set`, and 2 for mappings like `dict`.

At the moment, for my own project, I'm just looping through types of the same name in the various typing modules and getting the subscription sizes from there; most likely there's some significant overhead, but hopefully newer Python versions will be fast enough that it won't be noticeable! 😅

And again, thanks for the heads-up about the base `Union` matching all types! I'm just going to return `True` if the object is an annotation of any kind and false otherwise!

**Comment by leycec** [comment]
> I'm just looping through types of the same name in the various typing modules and getting the subscription sizes from there

*...heh.* Ye Ol' `_nparams` Trick, huh? The problem with that that's going away – probably by 2023, but possibly sooner. CPython devs *hate* most of the `typing` module, which they now consider legacy code that should just go away and die already. It's burdensome technical debt they'd rather not maintain. I get that... *sorta.*

Instead of violating privacy encapsulation on obsolete attributes, consider just hard-coding a mapping from type hint factories to subscription sizes. This is what @beartype does:

```python
# ....................{ PRIVATE ~ globals                  }....................
# Note that the builtin "range" class:
# * Initializer "range.__init__(start, stop)" is effectively instantiated as
#   [start, stop) -- that is to say, such that:
#   * The initial "start" integer is *INCLUSIVE* (i.e., the instantiated range
#     includes this integer).
#   * The final "stop" integer is *EXCLUSIVE* (i.e., the instantiated range
#     excludes this integer).
# * Publicizes these integers as the instance variables "start" and "stop" such
#   that these invariants are guaranteed:
#       >>> range(min, max).start == min
#       True
#       >>> range(min, max).stop == max
#       True

_ARGS_LEN_0 = range(0, 1)  # == [0, 1) == [0, 0]
'''
**Zero-argument length range** (i.e., :class:`range` instance effectively
equivalent to the integer ``0``, describing type hint factories subscriptable by
*no* child type hints).
'''


_ARGS_LEN_1 = range(1, 2)  # == [1, 2) == [1, 1]
'''
**One-argument length range** (i.e., :class:`range` instance effectively
equivalent to the integer ``1``, describing type hint factories subscriptable by
exactly one child type hint).
'''


_ARGS_LEN_2 = range(2, 3)  # == [2, 3) == [2, 2]
'''
**Two-argument length range** (i.e., :class:`range` instance effectively
equivalent to the integer ``2``, describing type hint factories subscriptable by
exactly two child type hints).
'''


_ARGS_LEN_3 = range(3, 4)  # == [3, 4) == [3, 3]
'''
**Three-argument length range** (i.e., :class:`range` instance effectively
equivalent to the integer ``3``, describing type hint factories subscriptable by
exactly three child type hints).
'''


_ARGS_LEN_1_OR_2 = range(1, 3)  # == [1, 3) == [1, 2]
'''
**One- or two-argument length range** (i.e., :class:`range` instance effectively
equivalent to the integer range ``[1, 2]``, describing type hint factories
subscriptable by either one or two child type hints).
'''

# ....................{ SIGNS ~ origin : args              }....................
# Fully initialized by the _init() function below.
HINT_SIGN_ORIGIN_ISINSTANCEABLE_TO_ARGS_LEN_RANGE: Dict[HintSign, range] = {
    # Type hint factories subscriptable by exactly one child type hint.
    HintSignAbstractSet: _ARGS_LEN_1,
    HintSignAsyncIterable: _ARGS_LEN_1,
    HintSignAsyncIterator: _ARGS_LEN_1,
    HintSignAwaitable: _ARGS_LEN_1,
    # HintSignByteString: _ARGS_LEN_1,
    HintSignCollection: _ARGS_LEN_1,
    HintSignContainer: _ARGS_LEN_1,
    HintSignCounter: _ARGS_LEN_1,
    HintSignDeque: _ARGS_LEN_1,
    HintSignFrozenSet: _ARGS_LEN_1,
    HintSignIterable: _ARGS_LEN_1,
    HintSignIterator: _ARGS_LEN_1,
    HintSignKeysView: _ARGS_LEN_1,
    HintSignList: _ARGS_LEN_1,
    HintSignMatch: _ARGS_LEN_1,
    HintSignMappingView: _ARGS_LEN_1,
    HintSignMutableSequence: _ARGS_LEN_1,
    HintSignMutableSet: _ARGS_LEN_1,
    HintSignPattern: _ARGS_LEN_1,
    HintSignReversible: _ARGS_LEN_1,
    HintSignSequence: _ARGS_LEN_1,
    HintSignSet: _ARGS_LEN_1,
    HintSignType: _ARGS_LEN_1,
    HintSignValuesView: _ARGS_LEN_1,

    # Type hint factories subscriptable by exactly two child type hints.
    HintSignAsyncGenerator: _ARGS_LEN_2,
    HintSignChainMap: _ARGS_LEN_2,
    HintSignDefaultDict: _ARGS_LEN_2,
    HintSignDict: _ARGS_LEN_2,
    HintSignItemsView: _ARGS_LEN_2,
    HintSignMapping: _ARGS_LEN_2,
    HintSignMutableMapping: _ARGS_LEN_2,
    HintSignOrderedDict: _ARGS_LEN_2,
    HintSignTuple: _ARGS_LEN_2,

    # Type hint factories subscriptable by exactly three child type hints.
    HintSignCoroutine: _ARGS_LEN_3,
    HintSignGenerator: _ARGS_LEN_3,
}
'''
Dictionary mapping from each sign uniquely identifying a PEP-compliant type hint
factory originating from an **isinstanceable origin type** (i.e., isinstanceable
class such that *all* objects satisfying type hints created by subscripting this
factory are instances of this class) to this factory's **argument length range**
(i.e., :class:`range` instance describing the minimum and maximum number of
child type hints that may subscript this factory).
'''

# ....................{ PRIVATE ~ main                     }....................
def _init() -> None:
    '''
    Initialize this submodule.
    '''

    # Defer function-specific imports.
    from beartype._util.py.utilpyversion import IS_PYTHON_AT_LEAST_3_13

    # If the active Python interpreter targets Python >= 3.13...
    if IS_PYTHON_AT_LEAST_3_13:
        # Add all signs uniquely identifying one- or two-argument type hint
        # factories under Python >= 3.13, which generalized various one-argument
        # type hint factories to accept an additional optional child type hint
        # via "PEP 696 – Type Defaults for Type Parameters".
        HINT_SIGN_ORIGIN_ISINSTANCEABLE_TO_ARGS_LEN_RANGE.update({
            HintSignAsyncContextManager: _ARGS_LEN_1_OR_2,
            HintSignContextManager: _ARGS_LEN_1_OR_2,
        })
    # Else, the active Python interpreter targets Python <= 3.12. In this
    # case...
    else:
        # Add all signs uniquely identifying two-argument type hint factories
        # under Python <= 3.12.
        HINT_SIGN_ORIGIN_ISINSTANCEABLE_TO_ARGS_LEN_RANGE.update({
            HintSignAsyncContextManager: _ARGS_LEN_1,
            HintSignContextManager: _ARGS_LEN_1,
        })
```

As the above code suggests, the standard `contextlib.AsyncContextManager` and `contextlib.ContextManager` type hint factories accept a variable range of subscriptions conditionally depending on the current CPython version. All other type hint factories currently accept a fixed length of subscriptions. This is QA nightmare fuel.

Wherever you may roam, you really want to reduce your dependence on private attributes but *especially* on obsolete private attributes. They will go away. Guaranteed. CPython devs are merciless about this sort of thing. Purple Karen shrugs. :woman_shrugging:

**Comment by sylvorg** [comment]
Oh no, no siree, I do _not_ have the biological memory for that QA nightmare. 😹 I need an automated way of doing it, otherwise I'll keep running into bugs i have no idea how to even _begin_ to fix.

Do you know when the `typing` types are going away? And if the subscription sizes are going to be added to the official types? I can just recurse up and down `abc.Iterable`'s super- and subclasses and test the sizes that way, if need be. 😹

**Comment by leycec** [comment]
> I need an automated way of doing it...

Doesn't exist, sadly. Let us commiserate together in shared misery. `typing` is hell. :sob: 

> Do you know when the typing types are going away?

**Superb question!** Alas, the answer is: "I am dumb and no nothing." The best-case answer is 2023. CPython devs want this logic gone now, but they also don't want to enrage everyone still stuck with legacy type hints. The worst-case answer is later this year:

> The deprecated functionality may eventually be removed from the `typing` module. Removal will occur no sooner than Python 3.9’s end of life, scheduled for October 2025.

> And if the subscription sizes are going to be added to the official types?

*...heh.* Not a snowball's chance in Hell. CPython devs hate runtime typing. Hell, CPython devs hate typing in general. There are *still* no type hints in CPython's standard library. No way they'd ever add a public API for introspecting type hints to *anything* outside the `typing` module – which they're actively trying to deprecate, obsolete, and delete as fast as possible.

`typing` is sadness. Cry, emoji cat! :crying_cat_face:

**Comment by sylvorg** [comment]
So I'll have to scour the Python documentation for all the subscription sizes? 😭 That's going to take forever! ... And so will refactoring my tests! 😭

Ah, well... I can always use the `typing` package on PyPI. 😹

**Comment by leycec** [comment]
> I can always use the `typing` package on PyPI.

**Ha-ha!** I had a similar (yet perhaps even more perverse) idea. Just:

* Embed a copy of the existing `typing` module inside your codebase *before* CPython 3.15 [or whatevahs] finally disembowels it.

The only downside there is a potential licensing conflict, depending on whatever license you've chosen. But... yeah. Other that licensing, that's *probably* the optimal solution. It's just a single file, so it's trivially embeddable. You control it's contents, so you're guaranteed that it will always satisfy your use case.

Oh, and... **Happy 2025!** Let us see if Canada can go a year without being violently invaded by the U.S. :partying_face:

**Comment by sylvorg** [comment]
> Oh, and... Happy 2025! 🥳

Sneaky, sneaky... 😏 It's a weird time to be a dual-citizen, that's for sure! 😹

But yeah, I think I might just do that! This is going in my base package, anyway! Does this license work with the MIT license, though?

"typing_extensions/LICENSE at main · python/typing_extensions · GitHub" https://github.com/python/typing_extensions/blob/main/LICENSE

The first footnote seems to imply it would...

**Comment by leycec** [comment]
> Does this license work with the MIT license, though?

Oh, wondrous day! You're using MIT too. I was concerned you might be tied to a GPL-like viral license. But MIT? No problem-o! MIT and BSD licensing is so permissive that it happily combines with basically everything.

Consider prefacing the existing `LICENSE` file in the root of your project with flimsy legalese that delineates the licensing: e.g.,

```
All files *EXCEPT* "{muh_package_name}/typing.py" are MIT-licensed.
"{muh_package_name}/typing.py" is licensed under the Python Software
Foundation License Version 2.
```

Something like that should happily suffice.

> It's a weird time to be a dual-citizen, that's for sure! 😹

*...ain't that truth.* I've been debating rescinding my U.S. citizenship for decades, but held back out of a general distrust of the federal government. That distrust has now proven to be well-founded. My disappointment is measurable in gigameters.

My wife's in a much more precarious situation. She only holds Canadian citizenship. Since she works for an American university, I've tried suggesting she apply for U.S. citizenship. She acknowledges that this now makes *rational* sense... but the *emotional* sense is still out to lunch. She can't bring herself to do it. I get that. Emotions beat up rationality and then steal rationality's lunch money.

This fragile construct we call "Canada" may only have a few more years of life left in it. We will miss you so much when you're gone, Canada. :canada:
